plugin/template/template.go

```diff
@@ -47,6 +48,15 @@ type templateData struct {
 	Type     string
 	Message  *dns.Msg
 	Question *dns.Question
+	ctx      context.Context
```

Comment:
Can we go forward without adding a context to a type? It should not be needed to have it here

Reply:
The problem is that this is fed into `text/template` so it needs all that state, and doesn't have the Context. The real answer would be for `text/template` to accept a `Context` but that's not gonna happen anytime soon. I need the context to call `metadata.ValueFunc(ctx, key)`. The actual metadata object can't be pulled out of the context because it is keyed by a private type from `metadata`.

Reply:
Options:

1) As is
2) Add a function in `metadata` that will return the map[string]func from the Context, and store that in the struct.

Reply:
I implemented #2 here.